File: 304-range-sum-query-2d-immutable/range-sum-query-2d-immutable.rs
struct NumMatrix {
    prefix_sums: Vec<Vec<i32>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumMatrix {

    fn new(matrix: Vec<Vec<i32>>) -> Self {
        let n_rows: usize = matrix.len();
        let n_cols: usize = if n_rows > 0 { matrix[0].len() } else { 0 };
        
        let mut prefix_sums = vec![vec![0; n_cols + 1]; n_rows + 1];
        for r in 0..n_rows{
            for c in 0..n_cols{
                 prefix_sums[r + 1][c + 1] = matrix[r][c] + prefix_sums[r][c + 1] + prefix_sums[r + 1][c] - prefix_sums[r][c]; 
            }
        }
        
        NumMatrix{ prefix_sums }
    }
    
    fn sum_region(&self, row1: i32, col1: i32, row2: i32, col2: i32) -> i32 {
        let row1: usize = row1 as usize;
        let row2: usize = row2 as usize;
        let col1: usize = col1 as usize;
        let col2: usize = col2 as usize;
        return  self.prefix_sums[row2 + 1][col2 + 1] -  self.prefix_sums[row2 + 1][col1] -  self.prefix_sums[row1][col2 + 1] +  self.prefix_sums[row1][col1];

    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * let obj = NumMatrix::new(matrix);
 * let ret_1: i32 = obj.sum_region(row1, col1, row2, col2);
 */

File: 304-range-sum-query-2d-immutable/range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        R, C = len(matrix), len(matrix[0])

        self.prefArr = [[0] * (C + 1) for _ in range(R + 1)]

        for r in range(R):
            prefix = 0
            for c in range(C):
                prefix += matrix[r][c]
                above = self.prefArr[r][c + 1]
                self.prefArr[r + 1][c + 1] = prefix + above

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        r1, c1, r2, c2 = row1 + 1, col1 + 1, row2 + 1, col2 + 1

        total = self.prefArr[r2][c2]
        upper = self.prefArr[r1-1][c2]
        left = self.prefArr[r2][c1-1]

        area = total - upper - left + self.prefArr[r1-1][c1-1]
        return area


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
